src/emulator/apphosting/serve.ts

```diff
@@ -2,37 +2,55 @@
  * Start the App Hosting server.
  * @param options the Firebase CLI options.
  */
-import { isIPv4 } from "net";
-import { checkListenable } from "../portUtils";
-import { wrapSpawn } from "../../init/spawn";
+
+import { spawn } from "cross-spawn";
+import { discoverPackageManager } from "./utils";
 
 /**
  * Spins up a project locally by running the project's dev command.
  */
-// eslint-disable-next-line @typescript-eslint/no-explicit-any
-export async function start(options: any): Promise<{ port: number }> {
-  let port = options.port;
-  while (!(await availablePort(options.host, port))) {
-    port += 1;
-  }
 
-  serve(options, port);
+export async function start(): Promise<{ hostname: string; port: number }> {
+  const hostUrl = await serve();
 
-  return { port };
+  const { hostname, port } = new URL(hostUrl);
+  return { hostname, port: +port };
 }
 
-function availablePort(host: string, port: number): Promise<boolean> {
-  return checkListenable({
-    address: host,
-    port,
-    family: isIPv4(host) ? "IPv4" : "IPv6",
+async function serve(): Promise<string> {
+  const rootDir = process.cwd();
+  const packageManager = await discoverPackageManager(rootDir);
+
+  /**
+   *  TODO: Should add a timeout here. If the hostUrl cannot be retrieved
+   *  in a timely manner, user is probably using an unsupported framework.
+   * */
+
+  const hostUrl = new Promise<string>((resolve, reject) => {
+    const serve = spawn(packageManager, ["run", "dev"], {
+      cwd: rootDir,
+    });
+
+    serve.stdout.on("data", (data: any) => {
+      process.stdout.write(data);
+      const url = getHostUrlFromString(data.toString());
```

Comment:
I think we can reasonably expect that the

1) dev server will run on localhost
2) dev server will respect the `PORT` environment variable

I feel like these expectation is probably as reasonable as expecting the dev server to print out the URL. wdyt?

Reply:
+1 reading from STDOUT definitely does not feel right, and I think these assumptions are valid.